kernel/src/metaptr.rs

```diff
@@ -0,0 +1,113 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Google LLC 2024.
+
+//! Defines the MetaPtr type
+
+use core::fmt::{Formatter, LowerHex, UpperHex};
+use core::ops::AddAssign;
+
+/// A pointer with target specific metadata.
+/// This should be used any time the kernel wishes to grant authority to the user, or any time
+/// the user should be required to prove validity of a pointer.
+#[derive(Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+#[repr(transparent)]
```

Comment:
Pulling this comment out from the conversation on the `MetaPtr` definition above to a separate thread:
> a value that refers to an "authenticated" or "valid" (we need general language for this) user accessible address... **On most architectures this will be a thin wrapper around a usize**... In CHERI, it will actually be a CHERI pointer, which is larger than a usize in hardware...

If I understand things correctly, with the current approach we will only track and enforce this pointer metadata on CHERI platforms, and for non-CHERI this implementation is a zero-overhead reduction to the pointers we were already passing around.

Would there be value instead in making this a "fat" pointer on non-CHERI platforms, i.e., explicitly adding the equivalent CHERI metadata to the `MetaPtr` struct? This comes with overhead obviously, but in theory at least the checks provided by CHERI are providing defense-in-depth security benefit; is that something that Tock should try to enable for non-CHERI platforms?

This may be an ignorant / infeasible question, i.e. if there's some HW tracking that can't be replicated by software getters/setters for `MetaPtr` (or if such a change would just be far more invasive than I appreciate on just a few minutes thought), but I wanted to ask once in this early design stage if we should try to be designing this abstraction to support [possible future] CHERI-style security for legacy / non-CHERI hardware?

Reply:
I think the "fat" part could easily be added in software, with caveats:

The small one is that encoding / decoding is quite an expensive operation (it is effectively a floating point number).

The large one is that actually the metadata is not a usize, it is a usize + 1 bit. That bit is hidden from most load/store operations so you can't read/write it manually. Replicating that is very hard, and if we just have to trust the user that the pointer they pass is "valid", we are back to trusting software as much as we did previously. The utility is therefore very low to have the metadata without the validity bit. And the power of the validly bit is that it is tracked in hardware.